sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/sqs/SqsUnboundedReader.java

```diff
@@ -0,0 +1,154 @@
+import com.amazonaws.services.sqs.model.ReceiveMessageResult;
+import com.google.common.collect.Lists;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import java.util.Set;
+import org.apache.beam.sdk.io.UnboundedSource;
+import org.apache.beam.sdk.io.UnboundedSource.CheckpointMark;
+import org.joda.time.Instant;
+
+class SqsUnboundedReader extends UnboundedSource.UnboundedReader<Message> {
+
+  public static final int MAX_NUMBER_OF_MESSAGES = 10;
+  private final SqsUnboundedSource source;
+  private final AmazonSQS sqs;
+
+  private Message current;
+  private final Queue<Message> messagesNotYetRead;
+  private Set<String> receiptHandlesToDelete;
+
+  public SqsUnboundedReader(SqsUnboundedSource source, SqsCheckpointMark sqsCheckpointMark) {
+    this.source = source;
+    this.current = null;
+
+    this.messagesNotYetRead = new ArrayDeque<>();
+    receiptHandlesToDelete = new HashSet<>();
+
+    final SqsConfiguration sqsConfiguration = source.getSqsConfiguration();
+    sqs =
+        AmazonSQSClientBuilder.standard()
+            .withClientConfiguration(sqsConfiguration.getClientConfiguration())
+            .withCredentials(sqsConfiguration.getAwsCredentialsProvider())
+            .withRegion(sqsConfiguration.getAwsRegion())
+            .build();
+
+    if (sqsCheckpointMark != null) {
+      if (sqsCheckpointMark.getReceiptHandlesToDelete() != null) {
+        receiptHandlesToDelete.addAll(sqsCheckpointMark.getReceiptHandlesToDelete());
+      }
+    }
+  }
+
+  @Override
+  public Instant getWatermark() {
+    return Instant.now();
+  }
+
+  @Override
+  public Instant getCurrentTimestamp() throws NoSuchElementException {
+    if (current == null) {
+      throw new NoSuchElementException();
+    }
+    return Instant.parse(current.getAttributes().get("Timestamp"));
+  }
+
+  @Override
+  public Message getCurrent() throws NoSuchElementException {
+    if (current == null) {
+      throw new NoSuchElementException();
+    }
+    return current;
+  }
+
+  @Override
+  public CheckpointMark getCheckpointMark() {
+    List<String> snapshotReceiptHandlesToDelete = Lists.newArrayList(receiptHandlesToDelete);
+    return new SqsCheckpointMark(this, snapshotReceiptHandlesToDelete);
+  }
+
+  @Override
+  public SqsUnboundedSource getCurrentSource() {
+    return source;
+  }
+
+  @Override
+  public boolean start() {
+    return advance();
+  }
+
+  @Override
+  public boolean advance() {
+    if (messagesNotYetRead.isEmpty()) {
+      pull();
+    }
+
+    current = messagesNotYetRead.poll();
+    if (current == null) {
+      return false;
+    }
+
+    receiptHandlesToDelete.add(current.getReceiptHandle());
+    return true;
+  }
+
+  @Override
+  public void close() {}
+
+  void delete(String receiptHandle) {
+    sqs.deleteMessage(source.getRead().queueUrl(), receiptHandle);
```

Comment:
Thinking if we can have this method receiving a List of handles and then using `deleteMessageBatch`  to delete them (argh hope AWS does not limit by 10 this but anyway will be better). Again not a blocker, this can be done in a future JIRA but given that the use is mostly 'ack' when checkpointing maybe worth checking.

Reply:
Trying to understand also if the differences of Standard vs FIFO queues can have a consequence for checkpointing.

Reply:
AWS limits to 10, I vote to defer.